File: web/src/models/Note.ts
import { isUndefined } from 'lodash';
import { NoteInterval, NoteSuffix, NoteValue, ScaleDegree } from '../enums';

//  C           D           E     F           G           A           B
//        C#/Db       D#/Eb             F#/Gb       G#/Ab       A#/Bb
//  B#                      Fb    E#                                  Cb
//  0     1     2     3     4     5     6     7     8     9     10    11

enum NoteSuffixLabel {
  Sharp = '\u266f',
  Flat = '\u266d',
  DoubleSharp = '\ud834\udd2a',
  DoubleFlat = '\ud834\udd2b',
}

export class Note {
  public static NUM_NOTES = 12;

  public SuffixLabel: string;

  public Base: NoteValue;
  public Suffix: NoteSuffix;

  public Degree: ScaleDegree;

  public Interval: NoteInterval;

  constructor(
    value: NoteValue = NoteValue.C,
    suffix: NoteSuffix = NoteSuffix.Natural
  ) {
    this.Base = value % Note.NUM_NOTES;
    this.Suffix = suffix;
  }

  private ModifiedValue(): NoteValue {
    return this.Base + this.Suffix;
  }

  private suffixLabel(suffix: NoteSuffix, long = false) {
    switch (suffix) {
      case NoteSuffix.DoubleFlat:
        return long ? 'Double Flat' : `${NoteSuffixLabel.DoubleFlat}`;
      case NoteSuffix.Flat:
        return long ? 'Flat' : `${NoteSuffixLabel.Flat}`;
      case NoteSuffix.Natural:
        return long ? 'Natural' : '';
      case NoteSuffix.Sharp:
        return long ? 'Sharp' : `${NoteSuffixLabel.Sharp}`;
      case NoteSuffix.DoubleSharp:
        return long ? 'Double Sharp' : `${NoteSuffixLabel.DoubleSharp}`;
    }
  }

  get Label(): string {
    return NoteValue[this.Base]
      ? `${NoteValue[this.Base]}${this.suffixLabel(this.Suffix)}`
      : 'ERROR';
  }

  get LongLabel(): string {
    return NoteValue[this.Base]
      ? `${NoteValue[this.Base]} ${this.suffixLabel(this.Suffix, true)}`
      : 'ERROR';
  }

  get Value(): NoteValue {
    return this.ModifiedValue();
  }

  get Pitch(): number {
    return (this.Base + this.Suffix) % Note.NUM_NOTES;
  }

  public Equals(note: Note) {
    return !!note && this.Base == note.Base && this.Suffix == note.Suffix;
  }

  static A(): Note {
    return new Note(NoteValue.A);
  }

  static B(): Note {
    return new Note(NoteValue.B);
  }

  static C(): Note {
    return new Note(NoteValue.C);
  }

  static D(): Note {
    return new Note(NoteValue.D);
  }

  static E(): Note {
    return new Note(NoteValue.E);
  }

  static F(): Note {
    return new Note(NoteValue.F);
  }

  static G(): Note {
    return new this(NoteValue.G);
  }

  public Natural(): Note {
    this.Suffix = NoteSuffix.Natural;
    return this;
  }

  public Sharp(): Note {
    this.Suffix = NoteSuffix.Sharp;
    return this;
  }

  public DoubleSharp(): Note {
    this.Suffix = NoteSuffix.DoubleSharp;
    return this;
  }

  public Flat(): Note {
    this.Suffix = NoteSuffix.Flat;
    return this;
  }

  public DoubleFlat(): Note {
    this.Suffix = NoteSuffix.DoubleFlat;
    return this;
  }

  public HalfStepUp(): Note {
    let pitch = this.Pitch;
    let next: Note;

    if (isUndefined(NoteValue[(pitch + 1) % Note.NUM_NOTES])) {
      if (pitch != this.Base) {
        next = new Note(pitch, NoteSuffix.Sharp);
      } else {
        next = new Note(pitch + 2, NoteSuffix.Flat);
      }
    } else {
      if (pitch + 1 == this.Base) {
        next = new Note(pitch + 2, NoteSuffix.Flat);
      } else {
        next = new Note(pitch + 1);
      }
    }

    return next;
  }

  //  C           D           E     F           G           A           B
  //        C#/Db       D#/Eb             F#/Gb       G#/Ab       A#/Bb
  //  B#                      Fb    E#                                  Cb
  //  0     1     2     3     4     5     6     7     8     9     10    11

  public WholeStepUp(): Note {
    let pitch = this.Pitch;
    let next: Note;

    if (isUndefined(NoteValue[(pitch + 2) % Note.NUM_NOTES])) {
      if (pitch + 1 == this.Base) {
        next = new Note(pitch + 3, NoteSuffix.Flat);
      } else {
        next = new Note(pitch + 1, NoteSuffix.Sharp);
      }
    } else {
      if (pitch - this.Base > 0) {
        next = new Note(pitch + 1, NoteSuffix.Sharp);
      } else {
        next = new Note(pitch + 2);
      }
    }

    return next;
  }
}

File: api/neck/Models/Entity/Note.cs
using neck.Enums;
using neck.Interfaces;
using System;
using System.ComponentModel.DataAnnotations.Schema;

namespace neck.Models.Entity
{
	public class Note : DbEntity, ILabelled, IEquatable<Note>
    {
        public NoteValue Base { get; set; }

        public NoteSuffix Suffix { get; set; }

        [NotMapped]
        public int? Octave { get; set; }

        [NotMapped]
        public string Label => $"{Base}{SuffixSymbol(Suffix)}";

        [NotMapped]
        public string PlainLabel => $"{Base}{SuffixPlainSymbol(Suffix)}";

        [NotMapped]
        public string SuffixLabel => $"{SuffixSymbol(Suffix)}";

        [NotMapped]
        public int Pitch {
            get => ((int)Base + (int)Suffix + Notes.Count) % Notes.Count;
        }

        // The difference in pitch between two notes,
        //	one being the root note of a scale
        [NotMapped]
        public Interval? Interval { get; set; }

        // Refers to the position of a particular note
        //	on a scale realtive to the tonic
        [NotMapped]
        public ScaleDegree? Degree { get; set; }

        public Note() { }

        public Note(NoteValue value, NoteSuffix suffix = NoteSuffix.Natural, int? octave = null)
        {
            Base = value;
            Suffix = suffix;
            Octave = octave;
        }

        public Note Copy()
        {
            var note = new Note(Base, Suffix, Octave);
            note.Degree = Degree;
            note.Interval = Interval;
            return note;
        }

		#region IEquatable

		public override bool Equals(object obj)
        {
            return Equals(obj as Note);
        }

        public bool Equals(Note other)
        {
            return other != null &&
                   Base == other.Base &&
                   Suffix == other.Suffix;
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Base, Suffix);
        }

		#endregion

		private string SuffixSymbol(NoteSuffix suffix)
        {
            switch (suffix)
            {
                case NoteSuffix.Sharp:
                    return "\u266f";
                case NoteSuffix.DoubleSharp:
                    return "\ud834\udd2a";
                case NoteSuffix.Flat:
                    return "\u266d";
                case NoteSuffix.DoubleFlat:
                    return "\ud834\udd2b";
                default:
                    return string.Empty;
            }
        }

        private string SuffixPlainSymbol(NoteSuffix suffix)
        {
            switch (suffix)
            {
                case NoteSuffix.Sharp:
                    return "#";
                case NoteSuffix.DoubleSharp:
                    return "##";
                case NoteSuffix.Flat:
                    return "b";
                case NoteSuffix.DoubleFlat:
                    return "bb";
                default:
                    return string.Empty;
            }
        }

        #region Utilities

        private int stepValue(int value, int step)
        {
            return (value + step) % Notes.Count;
        }

        private NoteValue noteValue(int value, int step) => (NoteValue)stepValue(value, step);

        private bool noteValueExists(int value, int step) => Enum.IsDefined(typeof(NoteValue), stepValue(value, step));

        #endregion

        #region Calculate note intervals & accidentals

        // TODO: Add half/whole step down?
        public Note HalfStepDown()
        {
            throw new NotImplementedException();
        }

        public Note HalfStepUp()
        {
            int step = 1;

            NoteValue nextValue;
            NoteSuffix nextSuffix = NoteSuffix.Natural;

            if (noteValueExists(Pitch, step))
            {
                if (Suffix == NoteSuffix.Flat)
                {
                    nextSuffix = NoteSuffix.Flat;
                    step = 2;
                }
                else if (Suffix == NoteSuffix.DoubleSharp)
                {
                    nextSuffix = NoteSuffix.Sharp;
                    step = 2;
                }
            }
            else
            {
                if (Suffix == NoteSuffix.Sharp || Suffix == NoteSuffix.DoubleSharp)
                {
                    nextSuffix = NoteSuffix.Sharp;
                    step = 0;
                }
                else
                {
                    step = 2;
                    nextSuffix = NoteSuffix.Flat;
                }
            }

            if (!noteValueExists(Pitch, step))
            {
                nextValue = noteValue((int)Base, step);

                if (nextSuffix == NoteSuffix.Flat)
                {
                    nextSuffix = NoteSuffix.DoubleFlat;
                }
            }
            else
			{
                nextValue = noteValue(Pitch, step);
			}
            return new Note((NoteValue)nextValue, nextSuffix);
        }

        //  C           D           E     F           G           A           B   
        //        C#/Db       D#/Eb             F#/Gb       G#/Ab       A#/Bb
        //  B#                      Fb    E#                                  Cb
        //  0     1     2     3     4     5     6     7     8     9     10    11 

        public Note WholeStepDown()
        {
            throw new NotImplementedException();
        }

        public Note WholeStepUp()
        {
            var step = 2;

            NoteValue nextValue;
            NoteSuffix nextSuffix = NoteSuffix.Natural;

            // TODO: get rid of this
            var calcFromBase = false;

            if (noteValueExists(Pitch, step))
            {
                if (Suffix == NoteSuffix.Sharp)
                {
                    step = 1;
                    nextSuffix = NoteSuffix.Sharp;
                }
                else if (Suffix == NoteSuffix.DoubleFlat)
                {
                    step = 3;
                    if (!noteValueExists(Pitch, step))
					{
                        step = 4;
                        nextSuffix = NoteSuffix.DoubleFlat;
                    }
                    else
                    {
                        nextSuffix = NoteSuffix.Flat;
                    }
				}
				else if (Suffix == NoteSuffix.DoubleSharp)
                {
                    calcFromBase = true;
                }
            }
            else
            {
                if (Suffix == NoteSuffix.Flat)
                {
                    step = 3;
                    nextSuffix = NoteSuffix.Flat;
                }
                else
                {
                    step = 1;
                    nextSuffix = NoteSuffix.Sharp;
                }
            }

            if (!noteValueExists(Pitch, step) || calcFromBase)
            {
                nextValue = noteValue((int)Base, step);

                if (nextSuffix == NoteSuffix.Flat)
                {
                    nextSuffix = NoteSuffix.DoubleFlat;
                }
                else if (nextSuffix == NoteSuffix.Sharp || nextSuffix == NoteSuffix.Natural)
                {
                    nextSuffix = NoteSuffix.DoubleSharp;
                }
            }
            else
            {
                nextValue = noteValue(Pitch, step);
            }
            return new Note(nextValue, nextSuffix);
        }
        #endregion

        #region Natural Note static constructors
        public static Note A()
        {
            return new Note(NoteValue.A, NoteSuffix.Natural);
        }

        public static Note B()
        {
            return new Note(NoteValue.B, NoteSuffix.Natural);
        }

        public static Note C()
        {
            return new Note(NoteValue.C, NoteSuffix.Natural);
        }

        public static Note D()
        {
            return new Note(NoteValue.D, NoteSuffix.Natural);
        }

        public static Note E()
        {
            return new Note(NoteValue.E, NoteSuffix.Natural);
        }

        public static Note F()
        {
            return new Note(NoteValue.F, NoteSuffix.Natural);
        }

        public static Note G()
        {
            return new Note(NoteValue.G, NoteSuffix.Natural);
        }
        #endregion
    }
}
